File: exodia/src/logger/logger.zig
const root = @This ();
const std = @import ("std");

const index = @import ("index");

const ansiterm = index.ansiterm;
const jdz = index.jdz;
const JdzGlobalAllocator = jdz.JdzGlobalAllocator (.{});
const termsize = index.termsize;

const Log = index.Log;
const Queue = index.Queue;
const Request = index.Request;
const Spin = index.Spin;
const Bar = index.Bar;

fn returnType (comptime name: [] const [] const u8) type
{
  var ptr = root;
  for (0 .. name.len - 1) |i| ptr = @field (ptr, name [i]);
  return (@typeInfo (@TypeOf (@field (ptr, name [name.len - 1]))).Fn.return_type orelse void);
}

pub const Logger = struct
{
  mutex: std.Thread.Mutex = .{},
  requests: Queue,
  spins: std.StringHashMap (Spin) = undefined,
  buffers: std.StringHashMap (Queue) = undefined,
  bar: Bar = undefined,
  allocator: *const std.mem.Allocator,
  looping: bool = true,
  cols: ?u16 = null,
  stderr_file: std.fs.File,
  stderr_writer: std.fs.File.Writer = undefined,
  buffered_stderr: std.io.BufferedWriter (4096, std.fs.File.Writer) = undefined,
  stderr: std.io.BufferedWriter (4096, std.fs.File.Writer).Writer = undefined,

  pub fn init (nproc: u32, allocator: *const std.mem.Allocator) !@This ()
  {
    var self: @This () = .{
      .requests    = Queue.init ("root"),
      .allocator   = allocator,
      .stderr_file = std.io.getStdErr (),
    };
    if (self.stderr_file.supportsAnsiEscapeCodes ()) self.cols = 0;
    self.stderr_writer = self.stderr_file.writer ();
    self.buffered_stderr = std.io.bufferedWriter (self.stderr_writer);
    self.stderr = self.buffered_stderr.writer ();
    self.spins = std.StringHashMap (Spin).init (self.allocator.*);
    self.buffers = std.StringHashMap (Queue).init (self.allocator.*);
    try self.spins.ensureTotalCapacity (nproc);
    try self.buffers.ensureTotalCapacity (nproc);
    return self;
  }

  pub fn deinit (self: *@This ()) void
  {
    self.looping = false;
  }

  pub fn enqueue (self: *@This (), request: Request) !void
  {
    self.mutex.lock ();
    defer self.mutex.unlock ();
    try self.requests.append (self.allocator, request);
  }

  pub fn enqueueTrace (self: *@This (), comptime name: [] const [] const u8, args: anytype) !returnType (name)
  {
    comptime var ptr = root;
    comptime var buf: [] const u8 = "";
    inline for (0 .. name.len - 1) |i|
    {
      buf = buf ++ name [i] ++ ".";
      ptr = @field (ptr, name [i]);
    }
    buf = buf ++ name [name.len - 1] ++ " (";
    inline for (0 .. args.len - 1) |i| buf = buf ++ std.fmt.comptimePrint ("{any}, ", .{args[i]});
    buf = buf ++ std.fmt.comptimePrint ("{any})", .{ args [args.len - 1], });
    try self.enqueue (.{ .kind = .{ .log = .TRACE, }, .data = .{ .message = buf, }, });
    return @call (.auto, @field (ptr, name [name.len - 1]), args);
  }

  fn dequeue (self: *@This ()) !bool
  {
    if (!self.mutex.tryLock ()) return false;
    defer self.mutex.unlock ();
    var log_rendered = false;
    while (self.requests.list.popFirst ()) |request|
    {
      log_rendered = try self.response (&request.data) or log_rendered;
      self.allocator.destroy (request);
    }
    return log_rendered;
  }

  fn logResponse (self: *@This (), request: *Request) !void
  {
    std.debug.assert (std.meta.activeTag (request.kind) == Request.Kind.log);
    std.debug.assert (std.meta.activeTag (request.data.?) == Request.Data.message);
    var log = Log.init (request);
    var looping = true;

    while (looping)
    {
      looping = try log.render (self);
      try self.animation ();
    }
  }

  fn spinResponse (self: *@This (), request: *Request) void
  {
    std.debug.assert (self.spins.count () < self.spins.capacity ());
    std.debug.assert (std.meta.activeTag (request.kind) == Request.Kind.spin);
    std.debug.assert (request.kind.spin.len > 0);
    std.debug.assert (!self.spins.contains (request.kind.spin));
    std.debug.assert (request.data != null);
    std.debug.assert (std.meta.activeTag (request.data.?) == Request.Data.message);
    self.spins.putAssumeCapacity (request.kind.spin, Spin.init (request.data.?.message));
  }

  fn killResponse (self: *@This (), request: *Request) void
  {
    std.debug.assert (std.meta.activeTag (request.kind) == Request.Kind.kill);
    std.debug.assert (request.kind.kill.len > 0);
    std.debug.assert (self.spins.remove (request.kind.kill));
  }

  fn bufferResponse (self: *@This (), request: *Request) !void
  {
    std.debug.assert (self.buffers.count () < self.buffers.capacity ());
    std.debug.assert (std.meta.activeTag (request.kind) == Request.Kind.buffer);
    std.debug.assert (request.kind.buffer.len > 0);
    if (!self.buffers.contains (request.kind.buffer))
      self.buffers.putAssumeCapacity (request.kind.buffer, Queue.init (request.kind.buffer));
    try self.buffers.getPtr (request.kind.buffer).?.append (self.allocator, request.*);
  }

  fn flushResponse (self: *@This (), request: *Request) void
  {
    std.debug.assert (self.buffers.count () > 0);
    std.debug.assert (std.meta.activeTag (request.kind) == Request.Kind.flush);
    std.debug.assert (request.kind.flush.len > 0);
    std.debug.assert (self.buffers.contains (request.kind.flush));
    while (self.buffers.getPtr (request.kind.flush).?.list.pop ()) |node| self.requests.list.prepend (node);
  }

  fn barResponse (self: *@This (), request: *Request) void
  {
    std.debug.assert (std.meta.activeTag (request.kind) == Request.Kind.bar);
    std.debug.assert (request.data != null);
    std.debug.assert (std.meta.activeTag (request.data.?) == Request.Data.max);
    std.debug.assert (request.data.?.max > 0);
    self.bar = Bar.init (request.data.?.max);
  }

  fn progressResponse (self: *@This ()) void
  {
    self.bar.incr ();
  }

  fn response (self: *@This (), request: *Request) !bool
  {
    var log_rendered = false;
    switch (request.kind)
    {
      .buffer   => try self.bufferResponse (request),
      .flush    => self.flushResponse (request),
      .spin     => self.spinResponse (request),
      .kill     => self.killResponse (request),
      .bar      => self.barResponse (request),
      .progress => self.progressResponse (),
      .log      => {
                     try self.logResponse (request);
                     log_rendered = true;
                   },
    }
    return log_rendered;
  }

  fn animation (self: *@This ()) !void
  {
    var spin_rendered = false;
    var it = self.spins.keyIterator ();
    while (it.next ()) |key|
    {
      try self.spins.getPtr (key.*).?.render (self, !spin_rendered);
      spin_rendered = true;
    }
    const bar_rendered = try self.bar.render (self, !spin_rendered);
    try self.clearFromCursorToScreenEnd ();
    for (0 .. self.spins.count ()) |i|
      if (i == 0) try self.cursorStartLine () else try self.cursorPreviousLine ();
    if (bar_rendered) for (0 .. 3) |i|
      if (!spin_rendered and i == 0) try self.cursorStartLine ()
      else try self.cursorPreviousLine ();
    try self.buffered_stderr.flush (); // don't forget to flush!
  }

  fn updateCols (self: *@This ()) !void
  {
    if (self.cols != null) self.cols = (try termsize.termSize (self.stderr_file)).?.width;
  }

  pub fn loop (self: *@This ()) !void
  {
    defer
    {
      self.spins.deinit ();
      self.buffers.deinit ();
      JdzGlobalAllocator.deinitThread ();
    }
    try self.hideCursor ();
    while (self.looping or self.requests.list.len > 0 or self.bar.running
      or self.spins.count () > 0 or self.buffers.count () > 0)
    {
      try self.updateCols ();
      if (!try self.dequeue ()) try self.animation ();
    }
    try self.clearFromCursorToScreenEnd ();
    try self.showCursor ();
    try self.buffered_stderr.flush (); // don't forget to flush!
  }

  pub fn clearFromCursorToLineEnd (self: @This ()) !void
  {
    if (self.cols != null) try ansiterm.clear.clearFromCursorToLineEnd (self.stderr);
  }

  fn clearFromCursorToScreenEnd (self: @This ()) !void
  {
    if (self.cols != null) try ansiterm.clear.clearFromCursorToScreenEnd (self.stderr);
  }

  fn showCursor (self: @This ()) !void
  {
    if (self.cols != null) try ansiterm.cursor.showCursor (self.stderr);
  }

  fn hideCursor (self: @This ()) !void
  {
    if (self.cols != null) try ansiterm.cursor.hideCursor (self.stderr);
  }

  fn cursorStartLine (self: @This ()) !void
  {
    if (self.cols != null) try ansiterm.cursor.setCursorColumn (self.stderr, 0);
  }

  fn cursorPreviousLine (self: @This ()) !void
  {
    if (self.cols != null) try ansiterm.cursor.cursorPreviousLine (self.stderr, 1);
  }

  pub fn updateStyle (self: @This (), style: ansiterm.style.Style) !void
  {
    if (self.cols != null) try ansiterm.format.updateStyle (self.stderr, style, null);
  }

  pub fn resetStyle (self: @This ()) !void
  {
    if (self.cols != null) try ansiterm.format.resetStyle (self.stderr);
  }

  pub fn writeAll (self: @This (), str: [] const u8) !void
  {
    try self.stderr.writeAll (str);
  }

  pub fn writeByte (self: @This (), byte: u8) !void
  {
    try self.stderr.writeByte (byte);
  }
};

File: exodia/go/logger/logger.go
package Logger

import (
  "bufio"
  "container/list"
  "errors"
  "math"
  "os"
  "sync"
  "time"
  "golang.org/x/term"
  "github.com/muesli/termenv"
  "github.com/tiawl/exodia/logger/color"
  "github.com/tiawl/exodia/logger/bar"
  "github.com/tiawl/exodia/logger/log"
  "github.com/tiawl/exodia/logger/queue"
  "github.com/tiawl/exodia/logger/spin"
  "github.com/tiawl/exodia/logger/request/bar"
  "github.com/tiawl/exodia/logger/request/progress"
  "github.com/tiawl/exodia/logger/request/spin"
  "github.com/tiawl/exodia/logger/request/kill"
  "github.com/tiawl/exodia/logger/request/log"
)

const (
  ClearCursorUntilScreenEndSeq = "J"
  CursorHorizontalSeq = "G"
)

type Type struct {
  nproc int
  mutex sync.Mutex
  requests *Queue.Type
  spins *list.List
  elements map [string] *list.Element
  bar *Bar.Type
  looping bool
  cols int
  writer *bufio.Writer
  output *termenv.Output
  restoreConsole func () error
}

func New (nproc int) *Type {
  var self *Type = &Type {
    nproc: nproc,
    mutex: sync.Mutex {},
    requests: Queue.New ("root"),
    elements: make (map [string] *list.Element, nproc),
    spins: list.New (),
    bar: Bar.New (0),
    looping: true,
    cols: -1,
    writer: bufio.NewWriter (os.Stderr),
  }
  var err error
  self.restoreConsole, err = termenv.EnableVirtualTerminalProcessing (termenv.DefaultOutput ())
  if err != nil { panic (err) }
  if term.IsTerminal (0) { self.cols = 0 }
  self.output = termenv.NewOutput (self.writer, termenv.WithProfile (termenv.ANSI256))
  return self
}

func (self *Type) Stop () {
  self.looping = false
}

func (self *Type) Enqueue (request interface {}) {
  self.mutex.Lock ()
  defer self.mutex.Unlock ()
  self.requests.Append (request)
}

func (self *Type) Log (message string, level string, color string) {
  var header termenv.Style = self.output.String (level)
  if (len (color) > 0) { header = header.Bold ().Foreground (self.output.Color (color)) }
  self.Enqueue (LogRequest.New (header.String (), message))
}

func (self *Type) Error (message string) {
  self.Log (message, "ERROR", Color.Red)
}

func (self *Type) Warn (message string) {
  self.Log (message, " WARN", Color.Yellow)
}

func (self *Type) Info (message string) {
  self.Log (message, " INFO", Color.Green)
}

func (self *Type) Note (message string) {
  self.Log (message, " NOTE", Color.Cyan)
}

func (self *Type) Debug (message string) {
  self.Log (message, "DEBUG", Color.Blue)
}

func (self *Type) Trace (message string) {
  self.Log (message, "TRACE", Color.Purple)
}

func (self *Type) Verb (message string) {
  self.Log (message, " VERB", Color.Pink)
}

func (self *Type) Raw (message string) {
  self.Log (message, "", "")
}

func (self *Type) Dequeue () bool {
  if !self.mutex.TryLock () { return false }
  defer self.mutex.Unlock ()
  var log_rendered bool = false
  for self.requests.Len () > 0 { log_rendered = self.Response (self.requests.PopFront ()) }
  return log_rendered
}

func (self *Type) BarResponse (request *BarRequest.Type) {
  self.bar = Bar.New (request.Max ())
}

func (self *Type) SpinResponse (request *SpinRequest.Type) *list.Element {
  // TODO: check Length before insert
  return self.spins.PushBack (Spin.New (request.Message ()))
}

func (self *Type) KillResponse (request *KillRequest.Type) {
  _ = self.spins.Remove (self.elements [request.Id ()])
  delete (self.elements, request.Id ())
}

func (self *Type) LogRender (log *Log.Type) bool {
  var entry string
  var looping bool
  entry, looping = log.Render (self.cols)
  self.WriteString (entry + "\n")
  return looping
}

func (self *Type) SpinRender (spin *Spin.Type, first bool) {
  if self.cols < 0 { return }
  if !first { self.WriteByte ('\n') }
  var chrono, pattern, message string
  var elapsed int64
  chrono, elapsed = spin.Chrono ()
  chrono = self.output.String (chrono).Bold ().Foreground (self.output.Color (Color.White)).String ()
  pattern = self.output.String (spin.Pattern (elapsed)).Bold ().Foreground (self.output.Color (spin.Color (elapsed))).String ()
  message = self.output.String (spin.Message ()).Bold ().Foreground (self.output.Color (Color.White)).String ()
  self.WriteString (chrono + pattern + message)
  self.output.ClearLineRight ()
}

func (self *Type) BarRender (first bool) bool {
  if self.cols < 0 { return false }
  var term_max int = (self.cols - 6) * 8
  self.bar.Update (term_max)
  if !self.bar.Running () { return false }

  var term_progress int = (self.bar.Progress () * term_max) / self.bar.Max ()

  var now time.Time = time.Now ()
  var ns int64 = int64 (now.Sub ((*self.bar.Last ())) % time.Second)
  if self.bar.Cursor () < term_progress && ns > 10_000_000 {
    var gap float64 = float64 ((term_progress - self.bar.Cursor ()) / 10)
    var log float64 = math.Log2 (math.Max (gap, 2)) - 1
    var shift int = int (math.Min (gap, log))
    self.bar.SetCursor (self.bar.Cursor () + (1 << shift))
    self.bar.SetLast (&now)
  }

  self.WriteString (self.output.String (self.bar.Top (self.cols, first)).Bold ().Foreground (self.output.Color (Color.White)).String ())
  self.output.ClearLineRight ()

  var percent int = (self.bar.Cursor () * 100) / term_max
  var color termenv.Color = self.output.Color (self.bar.Color (percent))
  self.WriteString (self.output.String (self.bar.MiddleStart ()).Bold ().Foreground (self.output.Color (Color.White)).String ())
  self.WriteString (self.output.String (self.bar.MiddleFilled ()).Bold ().Background (color).String ())
  self.WriteString (self.output.String (self.bar.MiddleEmpty (term_max)).Bold ().Foreground (color).String ())
  self.WriteString (self.output.String (self.bar.MiddleEnd (percent)).Bold ().Foreground (self.output.Color (Color.White)).String ())
  self.output.ClearLineRight ()

  self.WriteString (self.output.String (self.bar.Bottom (self.cols)).Bold ().Foreground (self.output.Color (Color.White)).String ())
  self.output.ClearLineRight ()

  return true
}

func (self *Type) LogResponse (request *LogRequest.Type) {
  var log *Log.Type = Log.New (request)
  var looping bool = true
  var err error

  for looping {
    looping = self.LogRender (log)
    err = self.writer.Flush ()
    if err != nil { panic (err) }
  }
}

func (self *Type) ProgressResponse (request *ProgressRequest.Type) {
  self.bar.Incr ()
}

func (self *Type) Response (request interface {}) bool {
  var log_rendered bool = false
  switch cast := request.(type) {
    case *SpinRequest.Type:
      self.elements [cast.Id ()] = self.SpinResponse (cast)
    case *KillRequest.Type:
      self.KillResponse (cast)
    case *BarRequest.Type:
      self.BarResponse (cast)
    case *ProgressRequest.Type:
      self.ProgressResponse (cast)
    case *LogRequest.Type:
      self.LogResponse (cast)
      log_rendered = true
    default:
      panic (errors.New ("Unknown Request type"))
  }
  return log_rendered
}

func (self *Type) UpdateCols () {
  if self.cols > -1 {
    var err error
    self.cols, _, err = term.GetSize (0)
    if err != nil { panic (err) }
  }
}

func (self *Type) WriteByte (b byte) {
  var err error = self.writer.WriteByte (b)
  if err != nil { panic (err) }
}

func (self *Type) WriteString (str string) {
  var err error
  _, err = self.writer.WriteString (str)
  if err != nil { panic (err) }
}

func (self *Type) CursorStartLine () {
  self.WriteString (termenv.CSI + "1" + CursorHorizontalSeq)
}

func (self *Type) ClearCursorUntilScreenEnd () {
  self.WriteString (termenv.CSI + ClearCursorUntilScreenEndSeq)
}

func (self *Type) GoUp (cond bool) {
  if cond { self.CursorStartLine ()
  } else { self.output.CursorPrevLine (1) }
}

func (self *Type) Animation () {
  var spin_rendered bool = false
  var err error
  for e := self.spins.Front (); e != nil; e = e.Next () {
    self.SpinRender (e.Value.(*Spin.Type), !spin_rendered)
    spin_rendered = true
  }
  var bar_rendered bool = self.BarRender (!spin_rendered)
  self.ClearCursorUntilScreenEnd ()
  for e := self.spins.Front (); e != nil; e = e.Next () {
    self.GoUp (e == self.spins.Front ())
  }
  if bar_rendered {
    for i := 0; i < 3; i++ {
      self.GoUp (i == 0 && !spin_rendered)
    }
  }
  err = self.writer.Flush ()
  if err != nil { panic (err) }
}

func (self *Type) Deferred () {
  var err error
  self.ClearCursorUntilScreenEnd ()
  self.output.ShowCursor ()
  err = self.restoreConsole ()
  if err != nil { panic (err) }
  err = self.writer.Flush ()
  if err != nil { panic (err) }
}

func (self *Type) Loop () {
  self.output.HideCursor ()
  defer self.Deferred ()
  for self.looping || self.requests.Len () > 0 || self.bar.Running () || self.spins.Len () > 0 {
    self.UpdateCols ()
    if !self.Dequeue () { self.Animation () }
  }
}
